tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -118,43 +282,118 @@ int main(int argc, char **argv)
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+        conn->server_protocol_version = S2N_TLS12;
+
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key(config, cert_chain_pem, private_key_pem));
+        EXPECT_SUCCESS(s2n_config_set_cache_store_callback(config, cache_store, session_cache));
+        EXPECT_SUCCESS(s2n_config_set_cache_retrieve_callback(config, cache_retrieve, session_cache));
+        EXPECT_SUCCESS(s2n_config_set_cache_delete_callback(config, cache_delete, session_cache));
+
+        EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+        /* Set up the connection to read from the fd */
+        EXPECT_SUCCESS(s2n_connection_set_read_fd(conn, client_to_server[0]));
+        EXPECT_SUCCESS(s2n_connection_set_write_fd(conn, server_to_client[1]));
+
+        /* Negotiate the handshake. */
+        EXPECT_SUCCESS(s2n_negotiate(conn, &blocked));
+
+        /* Make sure we did a abbreviated handshake */
+        EXPECT_TRUE(IS_RESUMPTION_HANDSHAKE(conn->handshake.handshake_type));
+
+        /* Ensure the message was delivered */
+        memset(buffer, 0, sizeof(buffer));
+        EXPECT_SUCCESS(bytes_read = s2n_recv(conn, buffer, sizeof(buffer), &blocked));
+        EXPECT_EQUAL(bytes_read, sizeof(MSG));
+        EXPECT_EQUAL(memcmp(buffer, MSG, sizeof(MSG)), 0);
+
+        /* Shutdown handshake */
+        shutdown_rc = -1;
+        do {
+            shutdown_rc = s2n_shutdown(conn, &blocked);
+            EXPECT_TRUE(shutdown_rc == 0 || (errno == EAGAIN && blocked));
+        } while(shutdown_rc != 0);
+
+        /* Clean up */
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+        EXPECT_SUCCESS(s2n_config_free(config));
+    }
+
+    /* Session resumption with bad session state on client side */
+    {
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+        conn->server_protocol_version = S2N_TLS12;
+
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key(config, cert_chain_pem, private_key_pem));
+        EXPECT_SUCCESS(s2n_config_set_cache_store_callback(config, cache_store, session_cache));
+        EXPECT_SUCCESS(s2n_config_set_cache_retrieve_callback(config, cache_retrieve, session_cache));
+        EXPECT_SUCCESS(s2n_config_set_cache_delete_callback(config, cache_delete, session_cache));
+
+        EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
 
-    EXPECT_SUCCESS(s2n_connection_free(conn));
+        /* Set up the connection to read from the fd */
+        EXPECT_SUCCESS(s2n_connection_set_read_fd(conn, client_to_server[0]));
+        EXPECT_SUCCESS(s2n_connection_set_write_fd(conn, server_to_client[1]));
+
+        /* Verify we failed to negotiate */
+        EXPECT_FAILURE(s2n_negotiate(conn, &blocked));
+
+        /* Clean up */
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+        EXPECT_SUCCESS(s2n_config_free(config));
+    }
```

Comment:
Do we need a test for falling back to full handshake if server doesn't recognize the session?

Reply:
Updated the initial handshake to happen with a session id that server doesn't recognize.